Sources/RDeviceIdentifier/RDeviceIdentifier.swift

```diff
@@ -0,0 +1,283 @@
+            // Try to find the device identifier in the keychain.
+            // Here we always have a bundle seed id.
+            var searchQuery = query
+            searchQuery[String(kSecMatchLimit)] = kSecMatchLimitOne
+            searchQuery[String(kSecReturnData)] = true
+            searchQuery[String(kSecAttrAccessGroup)] = accessGroup
+
+            var result: CFTypeRef?
+            let status = SecItemCopyMatching(searchQuery as CFDictionary, &result)
+            checkMissingAccessControl(status)
+
+            if status == errSecSuccess, let result = result as? Data {
+                return result
+            }
+            if status != errSecItemNotFound {
+                throw KeychainError.keychainLocked
+            }
+
+            return nil
+        }
+
+        func save(data: Data, for accessGroup: String) throws {
+            var saveQuery = query
+            saveQuery[String(kSecAttrAccessible)] = kSecAttrAccessibleWhenUnlocked
+            saveQuery[String(kSecValueData)] = data
+            saveQuery[String(kSecAttrAccessGroup)] = accessGroup
+
+            let status = SecItemAdd(saveQuery as CFDictionary, nil)
+            checkMissingAccessControl(status)
+            guard status == errSecSuccess else {
+                throw KeychainError.queryFailed(status: status)
+            }
+        }
+
+        func clear() {
+            SecItemDelete(query as CFDictionary)
+        }
+
+        private func checkMissingAccessControl(_ status: OSStatus?) {
+            // errSecNoAccessForItem is not defined for iOS, only OS X.
+            // Normally it would be found in <Security/SecBase.h>.
+            if status == errSecNoAccessForItem || status == errSecMissingEntitlement {
+                assertionFailure("Your application is lacking the proper keychain-access-group entitlements.")
+            }
+        }
+    }
+
+    enum KeychainError: Error, LocalizedError {
+        case keychainLocked
+        case queryFailed(status: OSStatus)
+        case internalError
+
+        var errorDescription: String? {
+            switch self {
+            case .keychainLocked: return "Keychain locked"
+            case .queryFailed(let status): return "Keychain query failed with code \(status)"
+            case .internalError: return "Internal error"
+            }
+        }
+    }
+
+    static var keychain = Keychain()
+
+    // MARK: - Cache
+
+    static var _uniqueDeviceIdentifier: String?
+
+    // MARK: - Helpers
+
+    private class func checkKeychainAccessGroupOrder(_ defaultAccessGroup: String) {
+        if let firstStopIndex = defaultAccessGroup.firstIndex(of: ".") {
+            let indexAfterStop = defaultAccessGroup.index(after: firstStopIndex)
+            let bundleId = String(defaultAccessGroup[indexAfterStop...])
+            if keychainAccessGroup == bundleId {
+                assertionFailure("\(keychainAccessGroup) is your default access group." +
```

Comment:
Can you throw an error instead of `assertionFailure`? Then this error can be caught in the caller.